### PR TITLE
Add Cabal analyzer

### PR DESCRIPTION
**Pull request by scruffystuffs**
Check comments for specific notes.  Other than those comments, this is pretty straightforward.

### CODE REVIEW

File: src/Strategy/Haskell/Cabal.hs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    getDepends mapB = pure . S.map PlanId . fromMaybe mempty $ M.lookup "depends" mapB
+
+installPlanDepends :: InstallPlan -> Set PlanId
+installPlanDepends InstallPlan {..} =
+  if planType == PreExisting
+    then mempty
+    else planComponents <> planDepends
+
+isDirectDep :: InstallPlan -> Bool
+isDirectDep InstallPlan {..} = planStyle == Just Local && planType == Configured
```
Comment (scruffystuffs) [inline src/Strategy/Haskell/Cabal.hs:86]: Not sure on the naming here.  When running against spectrometer, this is all of the plans created by the `spectrometer.cabal` file.  Examples are the spectrometer lib, the 3 binaries, and the unit-test binary.

File: src/Strategy/Haskell/Cabal.hs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  pure $ S.unions nested
+  where
+    getDepends :: Map Text (Set Text) -> Parser (Set PlanId)
+    getDepends mapB = pure . S.map PlanId . fromMaybe mempty $ M.lookup "depends" mapB
+
+installPlanDepends :: InstallPlan -> Set PlanId
+installPlanDepends InstallPlan {..} =
+  if planType == PreExisting
+    then mempty
+    else planComponents <> planDepends
```
Comment (scruffystuffs) [inline src/Strategy/Haskell/Cabal.hs:83]: We only care about a packages' deps if that package is not a base package.

File: src/Strategy/Haskell/Cabal.hs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+      <*> (obj .:? "components" >>= mergeComponents)
+
+mergeComponents :: Maybe (Map Text (Map Text (Set Text))) -> Parser (Set PlanId)
+mergeComponents Nothing = pure mempty
+mergeComponents (Just mapA) = do
+  nested <- traverse getDepends $ M.elems mapA
+  pure $ S.unions nested
+  where
+    getDepends :: Map Text (Set Text) -> Parser (Set PlanId)
+    getDepends mapB = pure . S.map PlanId . fromMaybe mempty $ M.lookup "depends" mapB
```
Comment (scruffystuffs) [inline src/Strategy/Haskell/Cabal.hs:77]: See [this example](https://github.com/fossas/spectrometer/blob/7268c12841d4f9ae8b0bf119650ea771a5ed7bfb/test/Haskell/testdata/cabal-plan.json#L33-L51) for why we need this Parser/Transformer.

File: src/Strategy/Haskell/Cabal.hs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    BuildPlan <$> obj .: "install-plan"
+
+instance FromJSON InstallPlan where
+  parseJSON = withObject "InstallPlan" $ \obj ->
+    InstallPlan <$> (obj .: "type" >>= parsePlanType)
+      <*> obj .: "id"
+      <*> obj .: "pkg-name"
+      <*> obj .: "pkg-version"
+      <*> (fromMaybe mempty <$> obj .:? "depends")
+      <*> (obj .:? "style" >>= parsePlanStyle)
```
Comment (cnr) [inline src/Strategy/Haskell/Cabal.hs:67]: this is good. we can improve it slightly with `traverse`:
```suggestion
      <*> (obj .:? "style" >>= traverse parsePlanStyle)
```

..where traverse has this type:
```hs
traverse :: (Text -> Parser PlanStyle) -> Maybe Text -> Parser (Maybe PlanStyle)
```
..which simplifies `parsePlanStyle`:

```hs
parsePlanStyle :: Text -> Parser PlanStyle
parsePlanStyle "global" = pure Global
parsePlanStyle "local" = pure Local
parsePlanStyle style = fail "unknown install plan style: " ++ T.unpack style
```

turns out `parsePlanStyle` has a more-general type, which GHC will tell you if you remove the type signature:
```hs
parsePlanStyle :: MonadFail m => Text -> m PlanStyle
```

File: src/Strategy/Haskell/Cabal.hs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+isDirectDep InstallPlan {..} = planStyle == Just Local && planType == Configured
+
+parsePlanStyle :: Maybe Text -> Parser (Maybe PlanStyle)
+parsePlanStyle Nothing = pure Nothing
+parsePlanStyle (Just style) = case T.toLower style of
+  "global" -> pure $ Just Global
+  "local" -> pure $ Just Local
+  _ -> fail $ "unknown install plan style" ++ T.unpack style
+
+parsePlanType :: Text -> Parser PlanType
```
Comment (cnr) [inline src/Strategy/Haskell/Cabal.hs:95]: [nit] this has a more-general type, `parsePlanType :: MonadFail m => Text -> m PlanType`. this makes it easier to see at a glance what the function does

File: src/Strategy/Haskell/Cabal.hs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    { cmdName = "cabal",
+      cmdArgs = ["v2-build", "--dry-run"],
+      cmdAllowErr = Never
+    }
+
+cabalPlanFilePath :: Path Rel File
+cabalPlanFilePath = $(mkRelFile "dist-newstyle/cache/plan.json")
+
+isCabalFile :: Path Abs File -> Bool
+isCabalFile file = is_v1 || is_v2
```
Comment (cnr) [inline src/Strategy/Haskell/Cabal.hs:113]: [nit] `is_v1`/`is_v2` might be confusing names to future-us; perhaps `isDotCabal` or `isCabalProject` or similar?

File: src/Strategy/Haskell/Cabal.hs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+doGraph plan = do
+  let parentId = planId plan
+  mapping parentId plan
+  for_ (installPlanDepends plan) $ \dep -> do
+    edge parentId dep
+    if isDirectDep plan
+      then direct dep
+      else pure ()
+
+shouldExclude :: InstallPlan -> Bool
```
Comment (cnr) [inline src/Strategy/Haskell/Cabal.hs:152]: `shouldExclude` seems to behave like a `shouldInclude` -- either rename or extract the `not` into the callsite below:

```hs
G.filter (not . shouldExclude) gr
```

File: src/Strategy/Haskell/Cabal.hs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+          dependenciesComplete = Complete
+        }
+
+doGraph :: Has (MappedGrapher PlanId InstallPlan) sig m => InstallPlan -> m ()
+doGraph plan = do
+  let parentId = planId plan
+  mapping parentId plan
+  for_ (installPlanDepends plan) $ \dep -> do
+    edge parentId dep
+    if isDirectDep plan
```
Comment (cnr) [inline src/Strategy/Haskell/Cabal.hs:148]: this is also called `when` from `Control.Monad`:
```hs
when (isDirectDep plan) $ direct dep
```

```hs
when :: Applicative f => Bool -> f () -> f ()
when True f = f
when False _ = pure ()
```

File: src/Strategy/Haskell/Cabal.hs (hunk 2 above)
Comment (scruffystuffs) [reply src/Strategy/Haskell/Cabal.hs:83]: Yeah good point.  This was left over from a previous filtering attempt.

**Pull request merged**